canopy/saplingjs/src/submitter.ts

```diff
@@ -0,0 +1,83 @@
+/**
+ * Copyright 2018-2020 Cargill Incorporated
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+interface BatchStatus {
+  statusType: string;
+  message: BatchMessage[];
+}
+
+interface BatchMessage {
+  transactionId: string;
+  errorMessage: string;
+  errorData: number[];
+}
+
+interface BatchInfo {
+  id: string;
+  status: BatchStatus;
+}
+
+export async function http(
+  method: string,
+  url: string,
+  data: Uint8Array,
+  headerFn: (request: XMLHttpRequest) => void
+): Promise<string> {
+  return new Promise((resolve, reject) => {
+    const request = new XMLHttpRequest();
+    request.open(method, url);
+    if (headerFn) {
+      headerFn(request);
+    }
+    request.onload = (): void => {
+      if (request.status >= 200 && request.status < 300) {
+        resolve(request.response);
+      } else if (request.status >= 400 && request.status < 500) {
+        reject(
+          Error('Failed to send request. Contact the administrator for help.')
+        );
+      } else {
+        reject(
+          Error(
+            'The server has encountered an error. Please contact the administrator.'
+          )
+        );
+      }
+    };
+    request.onerror = (): void => {
+      reject(
+        Error(
+          'The server has encountered an error. Please contact the administrator.'
+        )
+      );
+    };
+    request.send(data);
```

Comment:
What happens if the method doesn't take a payload? For instance if you are using this for a GET

Reply:
For `GET` and `HEAD` requests, the body/data argument is just ignored: https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/send

Reply:
But it is required in the function signature

Reply:
added none option